README.md
# md2 [![Build Status](https://travis-ci.com/3vilcookie/md2.svg?branch=master)](https://travis-ci.com/3vilcookie/md2)

Clone of the cracked, obsolete but easy to implement Message-Digest-2 hashing algorithm.

## Motivation

- Additional Task in my Information Security module
- Science and fun (no explanation needed)

## Installation
1. `git clone https://bitbucket.org/evilc00kie/md2`
2. cd md2
3. make

Tested on Windows 10 and Debian 8.

## Usage

Hash a single string

`echo -n "I feel so hashy 2day." | md2`
	
Hash whole file

`md2 debian.iso`

Test hashing function

`md2 -t`

Show help

`md2 -h`
## Reference

- RFC of the MD2-Algorithm with C Codeexample: http://www.faqs.org/rfcs/rfc1319.html
- Corrections of the RFC: https://www.rfc-editor.org/errata_search.php?rfc=1319
- Java implementation of the tool: https://www.gnu.org/software/gnu-crypto/ 


# License
GPL 


CHANGELOG.md
# Minor Release v0.1.0 (2020-04-19)
  * **Raphael Pour**
    * Import from BitBucket

*Released by Raphael Pour <info@raphaelpour.de>*


test/test.c
#include <stdlib.h>
#include <stdio.h>

#include "hash.h"

#ifndef BUILD_DATE
#define BUILD_DATE "fix your makefile"
#endif

#ifndef BUILD_VERSION
#define BUILD_VERSION "fix your makefile"
#endif

int main(int argc, char *argv[])
{
    printf("BuildDate: %s\n", BUILD_DATE);
    printf("BuildVersion: %s\n", BUILD_VERSION);

    int exitCode = EXIT_SUCCESS;

    printf("Test hash: ");
    const HASHTYPE *hash = md2hash( (const HASHTYPE*) "", 0);
    if ( memcmp( hash, testHash, BLOCK_LENGTH ) == 0 ) {
        puts("OK");
    }
    else
    {
        puts("NOT OK");
        exitCode = EXIT_FAILURE;
    }
    free( (void*)hash );

    return exitCode;
}




md2.h
/*
 * @file:	    md2.h
 * @brief:	    Generates a MD2 Hash of a given file
 * @date:	    19/04/2018
 * @author:	    Raphael Pour <s74020@informatik.htw-dresden.de>
 * @license:	GPL
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Make own types in order to not mix it up with others
typedef unsigned char HASHTYPE;
typedef size_t HASHLENGTH;

static HASHTYPE testHash[] = {//{{{
    0x83, 0x50, 0xe5, 0xa3, 
    0xe2, 0x4c, 0x15, 0x3d, 
    0xf2, 0x27, 0x5c, 0x9f,
    0x80, 0x69, 0x27, 0x73
};//}}}

static HASHTYPE piDigits[ 256 ] = {//{{{
  41, 46, 67, 201, 162, 216, 124, 1, 61, 54, 84, 161, 236, 240, 6,
  19, 98, 167, 5, 243, 192, 199, 115, 140, 152, 147, 43, 217, 188,
  76, 130, 202, 30, 155, 87, 60, 253, 212, 224, 22, 103, 66, 111, 24,
  138, 23, 229, 18, 190, 78, 196, 214, 218, 158, 222, 73, 160, 251,
  245, 142, 187, 47, 238, 122, 169, 104, 121, 145, 21, 178, 7, 63,
  148, 194, 16, 137, 11, 34, 95, 33, 128, 127, 93, 154, 90, 144, 50,
  39, 53, 62, 204, 231, 191, 247, 151, 3, 255, 25, 48, 179, 72, 165,
  181, 209, 215, 94, 146, 42, 172, 86, 170, 198, 79, 184, 56, 210,
  150, 164, 125, 182, 118, 252, 107, 226, 156, 116, 4, 241, 69, 157,
  112, 89, 100, 113, 135, 32, 134, 91, 207, 101, 230, 45, 168, 2, 27,
  96, 37, 173, 174, 176, 185, 246, 28, 70, 97, 105, 52, 64, 126, 15,
  85, 71, 163, 35, 221, 81, 175, 58, 195, 92, 249, 206, 186, 197,
  234, 38, 44, 83, 13, 110, 133, 40, 132, 9, 211, 223, 205, 244, 65,
  129, 77, 82, 106, 220, 55, 200, 108, 193, 171, 250, 36, 225, 123,
  8, 12, 189, 177, 74, 120, 136, 149, 139, 227, 99, 232, 109, 233,
  203, 213, 254, 59, 0, 29, 57, 242, 239, 183, 14, 102, 88, 208, 228,
  166, 119, 114, 248, 235, 117, 75, 10, 49, 68, 80, 180, 143, 237,
  31, 26, 219, 153, 141, 51, 159, 17, 131, 20
};//}}}

const unsigned short BLOCK_LENGTH = 16;
const unsigned BUFFER_LENGTH = 1024*10;
const unsigned short ITERATION_COUNT = 18;


const HASHTYPE* md2hash( const unsigned char *input, const HASHLENGTH length );
void dump( const HASHTYPE* buffer, const HASHLENGTH length );



source/hash.c
/*
 * @file:	    hash.c
 * @brief:	    Generates a MD2 Hash of a given file
 * @date:	    19/04/2018
 * @author:	    Raphael Pour <s74020@informatik.htw-dresden.de>
 * @license:	GPL
 */

#include "hash.h"

const HASHTYPE* md2hash( const HASHTYPE *input, const HASHLENGTH length )//{{{
{
    /*
     * Variables
     */
    // Counters for iterative processing
    HASHLENGTH i,j,k,t;

    // Lengths
    HASHLENGTH paddingLength = BLOCK_LENGTH - ( length % BLOCK_LENGTH );
    HASHLENGTH lengthWithPadding = length + paddingLength;
    HASHLENGTH lengthWithPaddingAndChecksum = length + paddingLength + BLOCK_LENGTH;

    // Checksum of the input data
    HASHTYPE checksum[ BLOCK_LENGTH ];
    
    // Storage for the processed data. This storage will be increased in size
    // during the hashing process.
    HASHTYPE *currentData;

    // Storage for the output
    HASHTYPE *output;

    // Temporary variables
    HASHTYPE c;
    HASHTYPE L = 0;


    /*
     * 1) Pad message so that its length is a multiple of 16
     */
    currentData = (HASHTYPE*) malloc( sizeof(HASHTYPE) * lengthWithPadding );

    if( length > 0 )
        memcpy( currentData, input, length );

    if( currentData == NULL )
    {
        printf( "Error on reallocating Memory for padded message" );
        exit( EXIT_FAILURE );
    }

    if( input )
        for( i=0; i<paddingLength; i++ )
            currentData[ length + i ] = paddingLength;

    /*
     * 2) Calculate Checksum
     */

    // Initialize checksum with zero
    for( i=0; i<BLOCK_LENGTH; i++)
        checksum[ i ] = 0;


    // Calculate Checksum block-by-block
    for( i=0; i<lengthWithPadding/BLOCK_LENGTH; i++ )
        for( j=0; j<BLOCK_LENGTH; j++ )
        {
            c = currentData[ i * BLOCK_LENGTH + j ];
            checksum[ j ] ^= piDigits[ c ^ L ];
            L = checksum[ j ]; 
        }
    
    currentData = (HASHTYPE*) realloc ( currentData, lengthWithPaddingAndChecksum );
    if( currentData == NULL )
    {
        printf( "Error on reallocating memory for checksum\n" );
        exit( EXIT_FAILURE );
    }

    // Append checksum to padded message
    memcpy( currentData + lengthWithPadding, checksum, BLOCK_LENGTH );


    /*
     * 3) Initialize 48-Byte Digest Buffer with zero
     */
    output = (HASHTYPE*) malloc( sizeof(HASHTYPE) * 48 );
    for( i=0; i<48; i++ )
        output[ i ] = 0;

    /*
     * 4) Process message in blocks
     */
    for( i=0; i<lengthWithPaddingAndChecksum/BLOCK_LENGTH; i++ )
    {
        // Leave the first Block of the output zero
        // Fill the second Block of the output with the first 16 Bytes of the
        // current data block
        //
        // Fill the third Block of the output with the second block of the
        // output XOR first block of the output
        for( j=0; j<BLOCK_LENGTH; j++ )
        {
            output[ BLOCK_LENGTH + j ] = currentData[ i * BLOCK_LENGTH + j ];
            output[ 2 * BLOCK_LENGTH + j ] = ( output[ BLOCK_LENGTH + j ] ^ output[ j ] );
        }

        t = 0;

        for( j=0; j<ITERATION_COUNT; j++ )
        {
            for( k=0; k<48; k++ )
            {
                output[ k ] ^= piDigits[ t ];
                t = output[ k ];
            }   
            t = ( t + j ) % 256;
        }
    }   

    free( currentData );

    output = realloc( output, BLOCK_LENGTH );

    return output;
}//}}}

void dump( const HASHTYPE* buffer, const HASHLENGTH length )//{{{
{
    HASHLENGTH i;
    for( i=0; i<length; i++ )
        printf( "%02x", buffer[ i ] );
}//}}}


include/hash.h
/*
 * @file:	    md2.h
 * @brief:	    Generates a MD2 Hash of a given file
 * @date:	    19/04/2018
 * @author:	    Raphael Pour <s74020@informatik.htw-dresden.de>
 * @license:	GPL
 */
#ifndef HASH_H
#define HASH_H

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Make own types in order to not mix it up with others
typedef unsigned char HASHTYPE;
typedef size_t HASHLENGTH;

static HASHTYPE testHash[] = {//{{{
    0x83, 0x50, 0xe5, 0xa3, 
    0xe2, 0x4c, 0x15, 0x3d, 
    0xf2, 0x27, 0x5c, 0x9f,
    0x80, 0x69, 0x27, 0x73
};//}}}

static HASHTYPE piDigits[ 256 ] = {//{{{
  41, 46, 67, 201, 162, 216, 124, 1, 61, 54, 84, 161, 236, 240, 6,
  19, 98, 167, 5, 243, 192, 199, 115, 140, 152, 147, 43, 217, 188,
  76, 130, 202, 30, 155, 87, 60, 253, 212, 224, 22, 103, 66, 111, 24,
  138, 23, 229, 18, 190, 78, 196, 214, 218, 158, 222, 73, 160, 251,
  245, 142, 187, 47, 238, 122, 169, 104, 121, 145, 21, 178, 7, 63,
  148, 194, 16, 137, 11, 34, 95, 33, 128, 127, 93, 154, 90, 144, 50,
  39, 53, 62, 204, 231, 191, 247, 151, 3, 255, 25, 48, 179, 72, 165,
  181, 209, 215, 94, 146, 42, 172, 86, 170, 198, 79, 184, 56, 210,
  150, 164, 125, 182, 118, 252, 107, 226, 156, 116, 4, 241, 69, 157,
  112, 89, 100, 113, 135, 32, 134, 91, 207, 101, 230, 45, 168, 2, 27,
  96, 37, 173, 174, 176, 185, 246, 28, 70, 97, 105, 52, 64, 126, 15,
  85, 71, 163, 35, 221, 81, 175, 58, 195, 92, 249, 206, 186, 197,
  234, 38, 44, 83, 13, 110, 133, 40, 132, 9, 211, 223, 205, 244, 65,
  129, 77, 82, 106, 220, 55, 200, 108, 193, 171, 250, 36, 225, 123,
  8, 12, 189, 177, 74, 120, 136, 149, 139, 227, 99, 232, 109, 233,
  203, 213, 254, 59, 0, 29, 57, 242, 239, 183, 14, 102, 88, 208, 228,
  166, 119, 114, 248, 235, 117, 75, 10, 49, 68, 80, 180, 143, 237,
  31, 26, 219, 153, 141, 51, 159, 17, 131, 20
};//}}}

#define BLOCK_LENGTH 16
#define BUFFER_LENGTH (1024*10)
#define ITERATION_COUNT 18


const HASHTYPE* md2hash( const unsigned char *input, const HASHLENGTH length );
void dump( const HASHTYPE* buffer, const HASHLENGTH length );

#endif // HASH_H


source/md2.c
/*
 * @file:	    md2.c
 * @brief:	    Generates a MD2 Hash of a given file
 * @date:	    19/04/2018
 * @author:	    Raphael Pour <s74020@informatik.htw-dresden.de>
 * @license:	GPL
 */

#include "hash.h"

int main( int argc, char* argv[] ) // {{{
{
    FILE *inputFile = stdin;

    HASHTYPE buffer[ BUFFER_LENGTH ];
    HASHTYPE *message = NULL;
    HASHLENGTH length = 0;
    HASHLENGTH read=0;

    // This programm is designed as a filter except a switch is provided.
    // No arguments means: read from stdin.
    // Filename as argument means: read from this file.
    if( argc == 2 )
    {
        // Tests hash algorithm by hashing "" and compare it with a hardcoded
        // one
        if( strcmp(argv[1],"-t") == 0 || strcmp(argv[1],"--test") ==0 )
        {
            printf( "md2(\"\"):" );
            const HASHTYPE *hash = md2hash( (const HASHTYPE*) "", 0);

            dump( hash, BLOCK_LENGTH );
            printf( " %s\n", ( memcmp( hash, testHash, BLOCK_LENGTH ) == 0 ) ? "OK" : "NOT OK" );

            free( (void*) hash );
            return EXIT_SUCCESS;
        }
        // Show help
        else if( strcmp(argv[1],"-h") == 0 || strcmp( argv[1],"--help") == 0 )
        {
            printf( "usage: %s [<filename>|-t|--test|-h|--help]\n", argv[0] );
            puts( "If no argument provided, md2 will read from stdin until EOF" );
            puts( "t ... Tests hashfunction by computing md2(\"\")" );
            puts( "h ... Prints help" );

            return EXIT_SUCCESS;
        }
        // Interpret argument as file and open it
        else
        {
            inputFile = fopen( argv[1], "rb" );
            if( inputFile == NULL )
            {
                perror( "Error opening input file" );
                return EXIT_FAILURE ;
            }
        }
    }

    /*
     * Copy Data from stream -file or stdin- into a buffer until EOF is reached
     */
    while( ! feof( inputFile ) )
    {
        read = fread( &buffer, 1, BUFFER_LENGTH, inputFile );

        if( read == 0 )
        {
            if( feof( inputFile ) )
                break;
            printf( "Error on reading file: %i\n", ferror( inputFile ) );
            return EXIT_FAILURE ;
        }

        message = realloc( message, length + read );
        if( message == NULL )
        {
            perror( "Error on allocating memory for input file" );
            return EXIT_FAILURE ;
        }

        memcpy( message + length, &buffer, read);
        length += read;
    }

    fclose( inputFile );

    // Generate hash and write it to stdout ('cause that's what filters do)
    dump( md2hash( message, length ), BLOCK_LENGTH );
    puts( "" );
    
    free( message );

    return EXIT_SUCCESS;
} //}}}
